panic: runtime error: index out of range

Hello Spencer,
this looks like an interesting algorithm. We're heavy users of the gorilla float compression algorithm (via the https://github.com/dgryski/go-tsz package) and I wondered how fpc compares.

To that end, I patched the eval program in the go-tsz library (i wrote this a while ago to get insights in compression performance across various types of input and lengths of the sequence being compressed)
like so:
```
diff --git a/eval/main.go b/eval/main.go
index f4ef015..1743fad 100644
--- a/eval/main.go
+++ b/eval/main.go
@@ -1,13 +1,16 @@
 package main
 
 import (
+       "bytes"
        "fmt"
-       "github.com/dgryski/go-tsz"
-       "github.com/dgryski/go-tsz/testdata"
+       //      "github.com/dgryski/go-tsz"
        "math"
        "math/rand"
        "os"
        "text/tabwriter"
+
+       "github.com/dgryski/go-tsz/testdata"
+       "github.com/spenczar/fpc"
 )
 
 // collection of 24h worth of minutely points, with different characteristics.
@@ -148,11 +151,16 @@ func main() {
        do := func(data []testdata.Point, comment string) string {
                str := ""
                for _, points := range intervals {
-                       s := tsz.New(data[0].T)
+                       buf := &bytes.Buffer{}
+                       w, err := fpc.NewWriterLevel(buf, 255)
+                       if err != nil {
+                               panic(err)
+                       }
                        for _, tt := range data[0:points] {
-                               s.Push(tt.T, tt.V)
+                               w.WriteFloat(tt.V)
                        }
-                       size := len(s.Bytes())
+                       w.Flush()
+                       size := buf.Len()
                        BPerPoint := float64(size) / float64(points)
                        str += fmt.Sprintf("\033[31m%d\033[39m\t%.2f\t", size, BPerPoint)
                }
```

and got the following panic:
```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/spenczar/fpc.(*fcm).predict(0xc4200143f0, 0x0)
	/home/dieter/go/src/github.com/spenczar/fpc/predictor.go:34 +0x38
github.com/spenczar/fpc.(*encoder).computeDiff(0xc420012300, 0x0, 0x1, 0xc42000e410)
	/home/dieter/go/src/github.com/spenczar/fpc/encode.go:198 +0x38
github.com/spenczar/fpc.(*encoder).encode(0xc420012300, 0x0, 0x0, 0xc42000e410, 0x1, 0x1, 0xc420041500)
	/home/dieter/go/src/github.com/spenczar/fpc/encode.go:218 +0x39
github.com/spenczar/fpc.(*blockEncoder).encode(0xc420052150, 0x0, 0x0, 0x0)
	/home/dieter/go/src/github.com/spenczar/fpc/encode.go:96 +0x6d
github.com/spenczar/fpc.(*Writer).writeUint64(0xc420014450, 0x0, 0x0, 0x0)
	/home/dieter/go/src/github.com/spenczar/fpc/writer.go:117 +0x5a
github.com/spenczar/fpc.(*Writer).writeFloat64(0xc420014450, 0x0, 0x0, 0x0)
	/home/dieter/go/src/github.com/spenczar/fpc/writer.go:110 +0x41
github.com/spenczar/fpc.(*Writer).WriteFloat(0xc420014450, 0x0, 0x0, 0x0)
	/home/dieter/go/src/github.com/spenczar/fpc/writer.go:70 +0x37
main.main.func1(0xc42008e000, 0x5a0, 0x5a0, 0x0, 0x0, 0xb9, 0x0)
	/home/dieter/go/src/github.com/dgryski/go-tsz/eval/main.go:160 +0x10e
main.main()
	/home/dieter/go/src/github.com/dgryski/go-tsz/eval/main.go:197 +0x187f
```

--- a/writer.go
+++ b/writer.go
@@ -10,7 +10,10 @@ import (
 
 const (
 	DefaultCompression = 10
-	MaxCompression     = 0xFF
+	// The reference implementation uses 255 for a max compression, but that
+	// hardly seems realistic: merely 32 will require 68 gigabytes of working
+	// memory to compute hashes. Beyond 35 we start hitting panics.
+	MaxCompression = 32
 
 	floatChunkSize = 8
 )

